arbnode/batch_poster.go

```diff
@@ -1468,11 +1530,61 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 			l1BoundMinTimestampWithBypass = arbmath.SaturatingUSub(timestampWithPadding, arbmath.BigToUintSaturating(maxTimeVariationDelaySeconds))
 		}
 	}
+	var lastPotentialMsg *arbostypes.MessageWithMetadata
+	var lastPotentialMsgPos arbutil.MessageIndex
+	var addMessageLoop func() bool
+	var getNextMessage func() (*arbostypes.MessageWithMetadata, error)
+	var breakLoopWhenErrorOccurs bool
 
-	for b.building.msgCount < msgCount {
-		msg, err := b.streamer.GetMessage(b.building.msgCount)
+	if b.espressoStreamer == nil {
+		msgCount, err := b.streamer.GetMessageCount()
 		if err != nil {
-			return false, fmt.Errorf("error getting message from streamer: %w", err)
+			log.Error("Error getting message count", "err", err)
+			return false, err
+		}
+		if msgCount <= batchPosition.MessageCount {
+			// There's nothing after the newest batch, therefore batch posting was not required
+			return false, nil
+		}
+
+		lastPotentialMsg, err = b.streamer.GetMessage(msgCount - 1)
+		if err != nil {
+			return false, err
+		}
+		addMessageLoop = func() bool { return b.building.msgCount < msgCount }
+		getNextMessage = func() (*arbostypes.MessageWithMetadata, error) {
+			msg, err := b.streamer.GetMessage(b.building.msgCount)
+			if err != nil {
+				return nil, err
+			}
+			return msg, nil
+		}
+		breakLoopWhenErrorOccurs = false
```

Comment:
why is this false? Shouldnt it always break look if the error occurs?

Reply:
There are two options:
- return directly
- break the loop

Breaking the loop means should continue the rest of the function.

Nitro's logic is to always fetch the `next` position, if errors occur, should retry again

For espresso streamer, when error encountering, we should break the loop to finish the batch